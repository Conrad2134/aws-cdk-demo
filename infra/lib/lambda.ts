import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as lambdaNode from '@aws-cdk/aws-lambda-nodejs';
import * as fs from 'fs';

export interface LambdaProps {
  entry: string;
  name: string;
}

export class Lambda extends cdk.Construct {
  public readonly handler: lambda.IFunction;

  constructor(scope: cdk.Construct, id: string, props: LambdaProps) {
    super(scope, id);

    // Make sure we can access the file, throw if we cannot.
    try {
      fs.accessSync(props.entry, fs.constants.F_OK);
    } catch {
      throw new Error(`Unable to access file at ${props.entry}`);
    }

    this.handler = new lambdaNode.NodejsFunction(this, `${id}StandardLambda`, {
      entry: props.entry,
      handler: 'handler',
      functionName: props.name,
      runtime: lambda.Runtime.NODEJS_14_X,
    });

    this.handler
      .metricErrors({
        period: cdk.Duration.minutes(1),
        statistic: cloudwatch.Statistic.SUM,
        unit: cloudwatch.Unit.COUNT,
      })
      .createAlarm(this, `${id}ErrorAlarm`, {
        alarmName: `${props.name}-error`,
        alarmDescription: `Alarm for errors in ${props.name}`,
        evaluationPeriods: 1,
        threshold: 1,
      });
  }
}
