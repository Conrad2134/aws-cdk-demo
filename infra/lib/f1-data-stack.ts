import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as path from 'path';
import { Lambda } from './lambda';

interface F1ServiceStackProps extends cdk.StackProps {
  stage: 'dev' | 'prod';
}

export class F1DataStack extends cdk.Stack {
  public readonly url: cdk.CfnOutput;
  public readonly stage: cdk.CfnOutput;

  constructor(scope: cdk.Construct, id: string, props: F1ServiceStackProps) {
    super(scope, id, props);

    const getDrivers = new Lambda(this, `${props.stage}-GetDrivers`, {
      name: `${props.stage}-get-drivers`,
      entry: path.join(__dirname, '..', '..', 'src', 'lambda', 'get-drivers.ts'),
    });

    const api = new apigw.RestApi(this, `${props.stage}-Api`);
    api.root.addResource('drivers').addMethod('GET', new apigw.LambdaIntegration(getDrivers.handler));

    this.stage = new cdk.CfnOutput(this, 'Stage', { value: props.stage });
    this.url = new cdk.CfnOutput(this, 'Url', {
      value: api.url,
    });
  }
}
