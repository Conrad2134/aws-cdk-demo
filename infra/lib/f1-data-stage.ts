import { CfnOutput, Construct, Stage, StageProps } from '@aws-cdk/core';
import { F1DataStack } from './f1-data-stack';

interface F1DataStageProps extends StageProps {
  stage: 'dev' | 'prod';
}

/**
 * Deployable unit of web service app
 */
export class F1DataStage extends Stage {
  public readonly url: CfnOutput;

  constructor(scope: Construct, id: string, props: F1DataStageProps) {
    super(scope, id, props);

    const service = new F1DataStack(this, 'F1DataStack', { stage: props.stage });

    // Expose F1DataStack's output one level higher
    this.url = service.url;
  }
}
