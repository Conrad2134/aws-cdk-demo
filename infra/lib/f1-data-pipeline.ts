import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
import { GitHubTrigger, ManualApprovalAction } from '@aws-cdk/aws-codepipeline-actions';
import { Construct, SecretValue, Stack, StackProps } from '@aws-cdk/core';
import { CdkPipeline, ShellScriptAction, SimpleSynthAction } from '@aws-cdk/pipelines';
import { F1DataStage } from './f1-data-stage';

/**
 * The stack that defines the application pipeline
 */
export class F1DataPipeline extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const sourceArtifact = new codepipeline.Artifact();
    const cloudAssemblyArtifact = new codepipeline.Artifact();

    const pipeline = new CdkPipeline(this, 'Pipeline', {
      // The pipeline name
      pipelineName: 'F1DataPipeline',
      cloudAssemblyArtifact,

      // Where the source can be found
      sourceAction: new codepipeline_actions.GitHubSourceAction({
        actionName: 'GitHub',
        output: sourceArtifact,
        oauthToken: SecretValue.secretsManager('github-token'),
        owner: 'Conrad2134',
        repo: 'aws-cdk-demo',
        branch: 'main',
        trigger: GitHubTrigger.POLL,
      }),

      // How it will be built and synthesized
      synthAction: SimpleSynthAction.standardNpmSynth({
        sourceArtifact,
        cloudAssemblyArtifact,
        installCommand: 'npm i -g npm@latest && npm ci',
        buildCommand: 'cd .. && npm ci',
        synthCommand: 'cd infra && npx cdk synth',
        subdirectory: 'infra',
      }),
    });

    // This is where we add the application stages
    const devStage = new F1DataStage(this, 'Dev', {
      env: { account: '194237782353', region: 'us-east-1' },
      stage: 'dev',
    });

    const devAppStage = pipeline.addApplicationStage(devStage);

    devAppStage.addActions(
      new ShellScriptAction({
        actionName: 'IntegrationTests',
        commands: ['npm i -g npm@latest', 'cd infra', 'npm ci', 'npm run test:integration'],
        additionalArtifacts: [sourceArtifact],
        runOrder: devAppStage.nextSequentialRunOrder(),
        useOutputs: {
          SERVICE_URL: pipeline.stackOutput(devStage.url),
        },
      }),
      new ManualApprovalAction({ actionName: 'Promote', runOrder: devAppStage.nextSequentialRunOrder() })
    );

    const prodStage = pipeline.addApplicationStage(
      new F1DataStage(this, 'Prod', {
        env: { account: '194237782353', region: 'us-east-1' },
        stage: 'prod',
      })
    );
  }
}
