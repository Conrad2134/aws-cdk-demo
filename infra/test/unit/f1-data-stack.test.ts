import { expect as expectCDK, haveOutput, haveResource, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { F1DataStack } from '../../lib/f1-data-stack';

test('outputs', () => {
  const app = new cdk.App();
  const stack = new F1DataStack(app, 'F1DataTestStack', { stage: 'dev' });

  expectCDK(stack).to(haveOutput({ outputName: 'Stage', outputValue: 'dev' }));
  expect(stack.url).not.toBeFalsy();
});

test('snapshot', () => {
  const app = new cdk.App();
  const stack = new F1DataStack(app, 'F1DataTestStack', { stage: 'dev' });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('key resources', () => {
  const app = new cdk.App();
  const stack = new F1DataStack(app, 'F1DataTestStack', { stage: 'dev' });

  // We have a lambda function
  expectCDK(stack).to(
    haveResource('AWS::Lambda::Function', {
      FunctionName: 'dev-get-drivers',
      Runtime: 'nodejs14.x',
    })
  );

  // We have an error alarm configured for that function
  expectCDK(stack).to(
    haveResource('AWS::CloudWatch::Alarm', {
      AlarmName: 'dev-get-drivers-error',
      MetricName: 'Errors',
      Statistic: 'Sum',
      Threshold: 1,
      Unit: 'Count',
    })
  );

  // Gateway and resource/method
  expectCDK(stack).to(haveResource('AWS::ApiGateway::RestApi'));

  expectCDK(stack).to(
    haveResource('AWS::ApiGateway::Resource', {
      PathPart: 'drivers',
      RestApiId: {
        Ref: 'devApiC3CC17DA',
      },
    })
  );

  expectCDK(stack).to(
    haveResource('AWS::ApiGateway::Method', {
      HttpMethod: 'GET',
      ResourceId: {
        Ref: 'devApidrivers15332EDF',
      },
    })
  );
});
