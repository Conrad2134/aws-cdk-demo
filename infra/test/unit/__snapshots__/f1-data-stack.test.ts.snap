// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Object {
  "Outputs": Object {
    "Stage": Object {
      "Value": "dev",
    },
    "Url": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "devApiC3CC17DA",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "devApiDeploymentStageprod52DF5A2B",
            },
            "/",
          ],
        ],
      },
    },
    "devApiEndpoint98C7470B": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "devApiC3CC17DA",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "devApiDeploymentStageprod52DF5A2B",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParametersdc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27ArtifactHashA9E4A5C0": Object {
      "Description": "Artifact hash for asset \\"dc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27\\"",
      "Type": "String",
    },
    "AssetParametersdc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27S3BucketA32DDD8B": Object {
      "Description": "S3 bucket for asset \\"dc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27\\"",
      "Type": "String",
    },
    "AssetParametersdc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27S3VersionKey5E2A4E82": Object {
      "Description": "S3 key for asset version \\"dc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "devApiAccount868FF95F": Object {
      "DependsOn": Array [
        "devApiC3CC17DA",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "devApiCloudWatchRole555CA2FB",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "devApiC3CC17DA": Object {
      "Properties": Object {
        "Name": "dev-Api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "devApiCloudWatchRole555CA2FB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "devApiDeployment810636E4254c537a5cd46d0ca84afdda704330b7": Object {
      "DependsOn": Array [
        "devApidriversGET3DB86A28",
        "devApidrivers15332EDF",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "devApiC3CC17DA",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "devApiDeploymentStageprod52DF5A2B": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "devApiDeployment810636E4254c537a5cd46d0ca84afdda704330b7",
        },
        "RestApiId": Object {
          "Ref": "devApiC3CC17DA",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "devApidrivers15332EDF": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "devApiC3CC17DA",
            "RootResourceId",
          ],
        },
        "PathPart": "drivers",
        "RestApiId": Object {
          "Ref": "devApiC3CC17DA",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "devApidriversGET3DB86A28": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "devGetDriversdevGetDriversStandardLambda36851C3E",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "devApidrivers15332EDF",
        },
        "RestApiId": Object {
          "Ref": "devApiC3CC17DA",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "devApidriversGETApiPermissionF1DataTestStackdevApi36D5F4CDGETdriversCDD8928B": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "devGetDriversdevGetDriversStandardLambda36851C3E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "devApiC3CC17DA",
              },
              "/",
              Object {
                "Ref": "devApiDeploymentStageprod52DF5A2B",
              },
              "/GET/drivers",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "devApidriversGETApiPermissionTestF1DataTestStackdevApi36D5F4CDGETdrivers63F99BEC": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "devGetDriversdevGetDriversStandardLambda36851C3E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "devApiC3CC17DA",
              },
              "/test-invoke-stage/GET/drivers",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "devGetDriversdevGetDriversErrorAlarmBD8DAFA0": Object {
      "Properties": Object {
        "AlarmDescription": "Alarm for errors in dev-get-drivers",
        "AlarmName": "dev-get-drivers-error",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "FunctionName",
            "Value": Object {
              "Ref": "devGetDriversdevGetDriversStandardLambda36851C3E",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "devGetDriversdevGetDriversStandardLambda36851C3E": Object {
      "DependsOn": Array [
        "devGetDriversdevGetDriversStandardLambdaServiceRoleCA98D48F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersdc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27S3BucketA32DDD8B",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27S3VersionKey5E2A4E82",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdc526fba3b959b607fb2889d3f2bfce36dcf6725762f9f1fb798e0f3e37cab27S3VersionKey5E2A4E82",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "dev-get-drivers",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "devGetDriversdevGetDriversStandardLambdaServiceRoleCA98D48F",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "devGetDriversdevGetDriversStandardLambdaServiceRoleCA98D48F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
